### ISSUE TITLE
redundant_closure lint causes a compile error when applied as is

### ISSUE THREAD

**Issue by asibahi** [rust-lang/rust-clippy#11604]
### Summary

Hello.

I have this function signature : 

```rust
fn traverse_inner(tree: TextTree, visit: &mut dyn FnMut(TextPiece))
```

Inside the function I call It recursively:

```rust
traverse_inner(tt, &mut |tp| visit(tp))
```

Clippy gives me redundant closure warning: 

```
warning: redundant closure
   --> src/helpers.rs:197:52
    |
197 |             .for_each(|tt| traverse_inner(tt, &mut |tp| visit(tp))),
    |                                                    ^^^^^^^^^^^^^^ help: replace the closure with the function itself: `&mut visit`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
    = note: `#[warn(clippy::redundant_closure)]` on by default
```

However this is a compile error since `visit` is not `mut`.

Full code with context is here: https://github.com/asibahi/mimiron/blob/765ecbd29bfc9c4c4796c5096111e8b70fe5df6a/src/helpers.rs#L190

### Lint Name

clippy::redundant_closure

### Reproducer

I tried this code:

```rust
#[allow(unused, unconditional_recursion)]
fn traverse_inner(num: u64, visit: &mut dyn FnMut(String)) {
    
    traverse_inner(num, &mut |i| visit(i));
    
}
```

I saw this happen:

```
warning: redundant closure
  --> src/main.rs:11:30
   |
11 |     traverse_inner(num, &mut |i| visit(i));
   |                              ^^^^^^^^^^^^ help: replace the closure with the function itself: `visit`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
   = note: `#[warn(clippy::redundant_closure)]` on by default
```

when applying the lint I get this compiler error 

```
error[E0596]: cannot borrow `visit` as mutable, as it is not declared as mutable
  --> src/main.rs:11:25
   |
11 |     traverse_inner(num, &mut visit);
   |                         ^^^^^^^^^^ cannot borrow as mutable
   |
note: the binding is already a mutable borrow
  --> src/main.rs:9:36
   |
9  | fn traverse_inner(num: u64, visit: &mut dyn FnMut(String)) {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^
help: try removing `&mut` here
   |
11 -     traverse_inner(num, &mut visit);
11 +     traverse_inner(num, visit);
   |
```

when applying the compiler's suggestion, however, the code compiles and runs.

### Version

```text
rustc 1.72.1 (d5c2e9c34 2023-09-13)
binary: rustc
commit-hash: d5c2e9c342b358556da91d61ed4133f6f50fc0c3
commit-date: 2023-09-13
host: aarch64-apple-darwin
release: 1.72.1
LLVM version: 16.0.5
```

**Comment by profetia** [comment]
I can no longer reproduce this in the latest nightly.

**Comment by ada4a** [comment]
> I can no longer reproduce this in the latest nightly.

me neither

@rustbot label S-fixed
